test/provisioner/provision_libvirt.go

```diff
@@ -179,36 +179,31 @@ func (l *LibvirtProvisioner) GetProperties(ctx context.Context, cfg *envconf.Con
 }
 
 func (l *LibvirtProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
-	// TODO: convert to use the libvirt.org/go/libvirt API.
-	//sPool, err := l.GetStoragePool()
-	//if err != nil {
-	//	return err
-	//}
-
-	//sVol, err := sPool.LookupStorageVolByName(l.volumeName)
-	//if err != nil {
-	//	return err
-	//}
-
-	//err = sVol.Upload(stream *Stream, 0, length uint64, libvirt.STORAGE_VOL_UPLOAD_SPARSE_STREAM)
-	//if err != nil {
-	//	return err
-	//}
-
-	//n, _ := sVol.GetName()
-
-	//fmt.Printf("%s\n", n)
-	cmd := exec.Command("/bin/bash", "-c",
-		fmt.Sprintf("virsh -c qemu:///system vol-upload --vol %s %s --pool default --sparse", l.volumeName, imagePath))
-	cmd.Dir = l.wd
-	cmd.Stdout = os.Stdout
-	// TODO: better handle stderr. Messages getting out of order.
-	cmd.Stderr = os.Stderr
-	err := cmd.Run()
+	sPool, err := l.GetStoragePool()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Stat(imagePath)
+	if err != nil {
+		return err
+	}
+	length := uint64(file.Size())
+
+	sVol, err := sPool.LookupStorageVolByName(l.volumeName)
+	if err != nil {
+		return err
+	}
+
+	stream, err := l.conn.NewStream(libvirt.STREAM_NONBLOCK)
```

Comment:
Hi @bookinabox , I was looking at the libvirt API. It seems we will need a blocking stream as well as cancel/abort/finish the stream properly. See https://libvirt.org/html/libvirt-libvirt-stream.html#virStreamNew

Reply:
Fixed and tested!